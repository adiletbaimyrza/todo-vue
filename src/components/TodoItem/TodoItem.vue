<template>
  <div @click="checkTodo" :class="$style['item']">
    <button
      @click.stop="checkTodo"
      :class="$style['item__button']"
      type="button"
    >
      <CoreIcon
        :src="
          isCompleted
            ? require('../../assets/circle-check.svg')
            : require('../../assets/circle.svg')
        "
      />
    </button>
    <p
      :class="[
        $style['item__content'],
        { [$style['item__content--completed']]: isCompleted },
      ]"
    >
      {{ content }}
    </p>
    <button
      @click.stop="deleteTodo"
      :class="[$style['item__button'], $style['item__button--delete']]"
      type="button"
    >
      <CoreIcon :src="require('../../assets/xmark.svg')" />
    </button>
  </div>
</template>

<script>
import CoreIcon from '../CoreIcon/CoreIcon.vue'

export default {
  name: 'TodoItem',
  components: {
    CoreIcon,
  },
  props: {
    id: {
      type: Number,
      required: true,
    },
    content: {
      type: String,
      required: true,
    },
    isCompleted: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    checkTodo() {
      this.$store.commit('check', this.id)
    },
    deleteTodo() {
      this.$store.commit('delete', this.id)
    },
  },
}
</script>

<style lang="scss" module>
.item {
  display: flex;
  width: 600px;
  padding: 1em 1.5em;
  align-items: center;
  color: var(--font-color-regular);
  background-color: var(--card-background-color);
  border-bottom: 1px solid var(--card-border-color);

  &:hover {
    background-color: var(--card-background-color--hover);
    cursor: pointer;

    .item__button--delete {
      visibility: visible;
    }
  }

  &__content {
    flex-grow: 1;
    text-align: left;
    padding: 0 1em;

    &--completed {
      color: #b1b1b1;
      text-decoration: line-through;
    }
  }

  &__icon {
    height: 2em;
    width: 2em;
  }

  &__button {
    &--delete {
      visibility: hidden;
    }
  }
}
</style>
