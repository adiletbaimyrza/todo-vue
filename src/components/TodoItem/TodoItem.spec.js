import { shallowMount } from '@vue/test-utils'
import TodoItem from './TodoItem.vue'

/* global jest describe it expect */
describe('TodoItem.vue', () => {
  const todoPropsNotCompleted = {
    id: 1,
    content: 'Test Todo',
    isCompleted: false,
  }
  const todoPropsCompleted = {
    id: 1,
    content: 'Test Todo',
    isCompleted: true,
  }

  it('renders correctly with required props', () => {
    const wrapper = shallowMount(TodoItem, {
      propsData: todoPropsNotCompleted,
    })
    expect(wrapper.exists()).toBe(true)
  })

  it('renders todo content correctly', () => {
    const wrapper = shallowMount(TodoItem, {
      propsData: todoPropsNotCompleted,
    })
    const content = wrapper.find('p')
    expect(content.text()).toBe('Test Todo')
  })

  it('calls checkTodo method when check button is clicked', async () => {
    const wrapper = shallowMount(TodoItem, {
      propsData: todoPropsNotCompleted,
    })
    const checkTodoMock = jest.fn()
    wrapper.vm.checkTodo = checkTodoMock
    const checkButton = wrapper.find('button')
    await checkButton.trigger('click')
    expect(checkTodoMock).toHaveBeenCalled()
  })

  it('calls deleteTodo method when delete button is clicked', async () => {
    const wrapper = shallowMount(TodoItem, {
      propsData: todoPropsNotCompleted,
    })
    const deleteTodoMock = jest.fn()
    wrapper.vm.deleteTodo = deleteTodoMock
    const deleteButton = wrapper.findAll('button').at(1)
    await deleteButton.trigger('click')
    expect(deleteTodoMock).toHaveBeenCalled()
  })

  it("applies 'item__content--completed' class if isCompleted === true", () => {
    const wrapper = shallowMount(TodoItem, {
      propsData: todoPropsCompleted,
    })
    const content = wrapper.find('p')
    expect(content.classes()).toContain(
      wrapper.vm.$style['item__content--completed']
    )
  })
})
