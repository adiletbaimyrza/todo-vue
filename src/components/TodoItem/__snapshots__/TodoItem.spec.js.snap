// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem.vue renders correctly with required props 1`] = `
<div
  class="item"
>
  <button
    class=""
    type="button"
  >
    <core-icon-stub
      height="24px"
      src=""
      width="24px"
    />
  </button>
  <p
    class="item__content"
  >
    Test Todo
  </p>
  <button
    class="item__button--delete"
    type="button"
  >
    <core-icon-stub
      height="24px"
      src=""
      width="24px"
    />
  </button>
</div>
`;
